### DISCUSSION TITLE
What is the definition of blobs?

### DISCUSSION THREAD

**Discussion by javierpagalo** [PMEAL/porespy#420] [Q&A] [answered]
I understand that the function `blobs` is like a way to get porous media with a control in the particles or something like that.
But is there any theory behind these? Is there a paper, book or scientific article related to the formulation?

I mean these lines:

```python
blobiness = np.array(blobiness)
shape = np.array(shape)
parallel = kwargs.pop('parallel', False)
divs = kwargs.pop('divs', 2)
cores = kwargs.pop('cores', None)
if np.size(shape) == 1:
    shape = np.full((3, ), int(shape))
sigma = np.mean(shape) / (40 * blobiness)
im = np.random.random(shape)
```

- Why do `sigma` is the mean of shape and it is divided by 40 the `blobiness`?
- What is the `blobiness`? 

I suppose it is a type of particle size, or maybe a shape factor, but I would like to know the exact meaning according to theory.

 Thank you.

**Comment by jgostick** [comment, marked as answer]
There is no science behind it.  It's just meant to provide image for prototyping.  The code is meant to provide a similar amount of blobs regardless of the image size, and that 40 is just a number that was chosen such that a `blobiness` of 1 gave reasonable sized blobs.  Looking at the code, as you did, is probably the best way to understand how it works.  Random numbers are generated between 0 and 1, a gaussian blur is performed using the `blobiness` to control the size and aspect ratio of the blurring kernel, and then the result is thresholded to give a boolean image with the correct porosity.

**Reply by javierpagalo** [reply]
Ok, it make sense. 
Thank you for that.
Btw, I would like to know how it affect to the pore size distribution.
Is there any relationship between blobiness and pore size distribution?

**Reply by jgostick** [reply]
No relationship that I know of.  The PSD of blobs is probably pretty narrow, and I'm guessing that changing the `blobiness` does affect the shape of the PSD, only the shift.  As I said, its useful prototyping tool, but probably not very scientific.  A more interesting 'blob' generator would be something like that generated using perlin noise instead of random noise.  Perlin is spatially correlated and fractal, so is more realistic.  You can generate blobs from perlin noise using `porespy.perlin_noise` but the api to that function is not good at the moment.  It doesn't work for a lot of size/frequency/octave combinations.  We're working on it.

**Reply by javierpagalo** [reply]
Ok, I understand.
Thank you for the answers and recommendations.
By the moment that's all.

**Reply by josephinoo** [reply]
What range of values does blobiness allow? From 0 to infinity? Or does it have any limits?.I await your prompt reply, thank you in advance.

**Reply by jgostick** [reply]
I would stick with 0.5 to 5.  It works for 2D images, so you can play with it as follows:

```python
import porespy as ps
import matplotlib.pyplot as plt
X = [1, 1]  # Choosing different values for each direction creates an anisotropic image
im = ps.generators.blobs(shape=[500, 500], porosity=0.5, blobiness=X)
plt.imshow(im, interpolation='none', origin='lower')
```

**Reply by josephinoo** [reply]
Ok, I understand.
Thank you for the answers and recommendations.

**Reply by javierpagalo** [reply]
Good Afternoon.
I have another question.
I have been reviewed some Authors. 
The method for the porous media generation in porespy.generators.blobs function seems to be based on the following algorithm presented by Jeffrey D. Hyman, C. Larrabee Winter, on their article "Stochastic generation of explicit pore structures by thresholding Gaussian random fields", page 3,  https://doi.org/10.1016/j.jcp.2014.07.046.
I'm doing a research and I just wanted to relate the blob function with another previous studies. 
Can you confirm if blob function is related on that algorithm?

Obviously is not equal, As you mention on the other answer, it is based on other parameters as blobiness and porosity.
But I mean, It could be a modification of that study. 

Waiting for your response
Thank You.

**Reply by jgostick** [reply]
I did not pull the blobs function from anywhere in particular...maybe a from a stack overflow answer (but it was a long time ago).  There is almost nothing of significance to the algorithm: generate random noise, apply Gaussian blur convolution, normalize and rescale the values, threshold, done.  Something this simple does not need a citation imho, although you should probably try to explain the steps in your paper (by looking at the code).
